shotover-proxy/src/transforms/cassandra/peers_rewrite.rs

```diff
@@ -0,0 +1,323 @@
+            .filter(|(_, m)| is_system_peers(*m))
+            .map(|(i, _)| i)
+            .collect::<Vec<_>>();
+
+        let mut response = message_wrapper.call_next_transform().await?;
+
+        for i in system_peers {
+            if let Some(new_port) = self.port {
+                rewrite_port(&mut response[i], new_port);
+            }
+
+            if self.emulate_single_node {
+                emulate_single_node(&mut response[i]);
+            }
+        }
+
+        Ok(response)
+    }
+}
+
+fn is_system_peers(message: &Message) -> bool {
+    if let Frame::Cassandra(_) = message.original {
+        if let Some(namespace) = message.namespace() {
+            if namespace.len() > 1 {
+                return namespace[0] == "system"
+                    && (namespace[1] == "peers" || namespace[1] == "peers_v2");
+            }
+        }
+    }
+
+    false
+}
+
+/// Emulate a single node by removing the rows from a query to system.peers(_v2)
+fn emulate_single_node(message: &mut Message) {
+    if let Frame::Cassandra(frame) = &message.original {
+        if let CassandraOperation::Result(CassandraResult::Rows { metadata, .. }) = &frame.operation
+        {
+            message.original = Frame::Cassandra(CassandraFrame {
+                version: frame.version,
+                stream_id: frame.stream_id,
+                tracing_id: frame.tracing_id,
+                warnings: frame.warnings.clone(),
+                operation: CassandraOperation::Result(CassandraResult::Rows {
+                    value: MessageValue::Rows(vec![]),
+                    metadata: metadata.clone(),
+                }),
+            });
+        } else {
+            panic!(
+                "Expected CassandraOperation::Result(CassandraResult::Rows), got {:?}",
+                frame
+            );
+        }
+    } else {
+        panic!("Expected Frame::Cassandra, got {:?}", message.original);
+    }
+}
+
+fn rewrite_port(message: &mut Message, new_port: u32) {
+    if let Frame::Cassandra(ref mut frame) = message.original {
+        if let CassandraOperation::Result(CassandraResult::Rows {
+            ref mut value,
+            ref metadata,
+        }) = frame.operation
+        {
+            let port_column_index = metadata
+                .col_specs
+                .iter()
+                .position(|col| col.name.as_str() == "native_port");
+
+            if let Some(i) = port_column_index {
+                if let MessageValue::Rows(ref mut rows) = *value {
+                    for row in rows.iter_mut() {
+                        row[i] = MessageValue::Integer(new_port as i64, IntSize::I32);
+                    }
+                }
+            }
+        } else {
+            panic!(
+                "Expected CassandraOperation::Result(CassandraResult::Rows), got {:?}",
+                frame
+            );
+        }
+    } else {
+        panic!("Expected Frame::Cassandra, got {:?}", message.original);
```

Comment:
I've been thinking about this a bit.
In general transforms are supposed to handle any message type but there are some transforms that only make sense with one message type (like this one) and would indicate a user configuration error if they encountered another message type.
A panic here is no good because it is possible for the user to hook a Redis Source up to a CassandraPeersRewrite and we shouldn't panic even under bad configuration.
We could return an Err here but that still allows the system to run under bad configuration.
The best solution here is to add a system to our validation system that allows transforms to define what message types they are compatible with.
Then we can fail to start shotover on a bad configuration.

However the transform invariants currently don't specify whether a transform instance will always receive messages of the same type.
If we cant think of a use case for multiple message types going down the same chain then we should:
* explicitly specify in the invariants that transforms will always receive messages of the same type and that transforms should never allow messages of mixed types to flow down the chain. (They can completely replace messages of type A with type B though)
* implement the allowed message type configuration validation
This should of course go in a separate issue so we can address it separately to this PR

For now I think the current implementation that panics is the best solution here.

Reply:
The panic here is because we should never be passing a Redis message into `rewrite_port` or `emulate_single` node unless the `is_system_peers` function is broken.

Reply:
Oh, true!

Maybe include that in the panic message, and ill dump my above thoughts in an issue.

Reply:
Yep, I will make it more obvious what is happening there.